File: 692.top-k-frequent-words.cpp
/*
 * @lc app=leetcode id=692 lang=cpp
 *
 * [692] Top K Frequent Words
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
typedef pair<string, int> wc;

class Solution {
public:
    // T: O(n * log(k))
    // S: O(n)
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> f;
        for (const string& w : words) {
            ++f[w];
        }

        // Sort by count DESC and then by lexicographical order ASC
        // Comparator symbols are reversed, because we fix the heap size at k and keep popping less frequent words
        auto cmp = [&](const wc& p1, const wc& p2) {
            return p1.second > p2.second || (p1.second == p2.second && p1.first < p2.first);
        };
        priority_queue<wc, vector<wc>, decltype(cmp)> min_heap(cmp);
        for (const auto& e : f) {
            min_heap.push(e);
            if (min_heap.size() > k) {
                min_heap.pop();
            }
        }

        // Results are written to the container reversely as heap
        // top is always the node with smallest freq. 
        vector<string> res(k, "");
        int p = k - 1;
        while (!min_heap.empty()) {
            res[p--] = min_heap.top().first; min_heap.pop();
        }
        return res;
    }
};
// @lc code=end

File: 692.top-k-frequent-words.py
#
# @lc app=leetcode id=692 lang=python3
#
# [692] Top K Frequent Words
#
import heapq
from collections import Counter
from typing import *

# @lc code=start
class Node:
    def __init__(self, word, freq):
        self.word = word
        self.freq = freq

    def __lt__(self, other):
        if self.freq == other.freq:
            return self.word > other.word
        return self.freq < other.freq


class Solution:
    # T: O(n * log(k))
    # S: O(n)
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        cnt = Counter(words)
        
        heap: List[Node] = []

        # To keep the space usage k, instead of simulating a max heap with 
        # elements (-c, w), define a node type with customized comparator,
        # such that heap top is the least element with the largest lexicographical
        # rank.
        for w, c in cnt.items():
            node = Node(w, c)
            if len(heap) == k:
                heapq.heappushpop(heap, node)
            else:
                heapq.heappush(heap, node)
        
        # Results are written to the container reversely as heap
        # top is always the node with smallest freq. 
        res = [""] * len(heap)
        for i in range(k - 1, -1, -1):
            res[i] = heapq.heappop(heap).word
        return res

# @lc code=end

